frame/undercollateralized-loans/src/helpers.rs

```diff
@@ -0,0 +1,529 @@
+		let destination = &loan_account_id;
+		let amount = *loan_config.collateral();
+		T::MultiCurrency::transfer(collateral_asset_id, source, destination, amount, keep_alive)?;
+		// Transfer borrowed assets from market's account to the borrower's account.
+		let borrow_asset_id = *loan_config.borrow_asset_id();
+		let source = loan_config.market_account_id();
+		let destination = &borrower_account_id;
+		let amount = *loan_config.principal();
+		T::MultiCurrency::transfer(borrow_asset_id, source, destination, amount, keep_alive)?;
+		// Mint 'principal' amount of debt tokens to the loan's account.
+		// We use these tokens to indicate which part of principal is not refunded yet.
+		let debt_token_id =
+			crate::DebtTokenForMarketStorage::<T>::get(loan_config.market_account_id())
+				.ok_or(Error::<T>::MarketDoesNotExist)?;
+		<T as Config>::MultiCurrency::mint_into(debt_token_id, &loan_account_id, amount)?;
+		// Loan is active now.
+		// Remove loan configuration from the non-activated loans accounts ids storage.
+		crate::NonActiveLoansStorage::<T>::remove(loan_account_id.clone());
+		// Build payment schedule.
+		for timestamp in loan_config.schedule().keys() {
+			crate::ScheduleStorage::<T>::mutate(timestamp, |loans_accounts_ids| {
+				loans_accounts_ids.insert(loan_account_id.clone())
+			});
+		}
+		Ok(loan_config)
+	}
+
+	// Repay any amount of money
+	pub(crate) fn do_repay(
+		payer_account_id: T::AccountId,
+		loan_account_id: T::AccountId,
+		repay_amount: T::Balance,
+		keep_alive: bool,
+	) -> Result<T::Balance, DispatchError> {
+		// Get loan's info.
+		let loan_config = Self::get_loan_config_via_account_id(&loan_account_id)?;
+		let borrow_asset_id = loan_config.borrow_asset_id();
+		// Transfer 'amount' of assets from the payer account to the loan account
+		T::MultiCurrency::transfer(
+			*borrow_asset_id,
+			&payer_account_id,
+			&loan_account_id,
+			repay_amount,
+			keep_alive,
+		)
+	}
+
+	pub(crate) fn do_liquidate(loan_config: &LoanConfigOf<T>) -> () {
+		Self::deposit_event(crate::Event::<T>::TheLoanWasSentToLiquidation {
+			loan_config: loan_config.clone(),
+		});
+	}
+
+	// Close loan contract since loan is paid.
+	pub(crate) fn close_loan_contract(loan_config: &LoanConfigOf<T>, keep_alive: bool) {
+		// Transfer collateral to borrower's account.
+		T::MultiCurrency::transfer(
+			*loan_config.collateral_asset_id(),
+			loan_config.market_account_id(),
+			loan_config.borrower_account_id(),
+			*loan_config.collateral(),
+			keep_alive,
+		)
+	    // May happens if borrower's account died for some reason.
+        // TODO: @mikolaichuk: what we are going to with such situations? 
+        // Can we transfer collateral at market account in this case?
+        // Perhaps we have to allow manager to transfer collateral to any account in such cases? 
+        .map_or_else(|error| log::error!("Collateral was not transferred back to the borrower's account due to the following error: {:?}", error), |_| ());
```

Comment:
Do you not want this function to fail of either of the transfers fail?

Reply:
@kevin-composable it means that not closed contract will stay in our storage forever, since we try to close it only once after it was paid.